### PR TITLE
[NEW] Custom User Status

### PR DESCRIPTION
**Pull request by Hudell**
This PR merges together the implementations of custom user status on two different PRs: #12710 (by @wreiske ) and #9353 (by me)

It adds the ability for admins to register additional user status and also for users to type in any custom message as their current status.

image

image

### CODE REVIEW

**Comment by engelgabriel** [comment]
Setting your Status Message should be independent form your Online Presence

**Comment by engelgabriel** [comment]
image

having the expiration is far more important than the online presence

**Comment by engelgabriel** [comment]
maybe you can choose the **Presence Status**, but if you leave it empty, the **Status Message** will be independent of the **Presence Status**, but if you choose a color, like Busy for a custom **Status Message** like "In a meeting", then you set both together.

**Comment by wreiske** [comment]
LEAK! LEAK!
image

Does this mean that this will be in 1.0.0?

**Comment by Hudell** [comment]
Huh, not sure who made that picture, but no :/
Everything is ready for merging here (unless the design team decides to change anything) but we're waiting until after 1.0, since this release is already pretty big.
I'm keeping track of this PR to make sure no conflict shows up again and I'll get someone to merge it as soon as 1.0 is out.

**Comment by wreiske** [comment]
It's on HTTP://Rocket.Chat

File: app/ui/client/components/selectDropdown.html [hunk 1]
```diff
@@ -1,19 +1,18 @@
...
+				</select>
 			</div>
 		</label>
-		{{#if open}}
-		<div class="fadeInDown">
-			{{ > popupList .}}
-		</div>
-		{{/if}}
 	</div>
 </template>
```
Comment (tassoevan) [inline app/ui/client/components/selectDropdown.html]: Not needed. (Wow, this is obsolete!)

File: app/ui/client/components/selectDropdown.js [hunk 2]
```diff
@@ -1,23 +0,0 @@
...
-	},
-});
-Template.selectDropdown.helpers({
-	open() {
-		return Template.instance().open.get();
-	},
-});
-Template.selectDropdown.onCreated(function() {
-	this.open = new ReactiveVar(false);
-});
```
Comment (tassoevan) [inline app/ui/client/components/selectDropdown.js]: Not needed. (Wow, this is obsolete!)

Comment (Hudell) [reply app/ui/client/components/selectDropdown.js]: The PR is removing this code, not adding it :P

File: app/ui-flextab/client/tabs/userInfo.js [hunk 3]
```diff
@@ -85,6 +87,15 @@ Template.userInfo.helpers({
 		return userStatus;
 	},
 
+	userStatusText() {
+		if (s.trim(this.statusText)) {
+			return emojione.render(s.escapeHTML(this.statusText));
```
Comment (wreiske) [inline app/ui-flextab/client/tabs/userInfo.js]: I added this code to allow emojis in status messages before I knew how emojis were integrated into Rocket.

Maybe take a look at https://github.com/RocketChat/Rocket.Chat/blob/d53e264b83ccc971236123853e4b68c9eb6beeb4/app/emoji/client/emojiParser.js#L26

Want to make sure custom emojis work!

**Comment by boblmartens** [comment]
Just wanted to jump in here and watch what is going on. We have a department that would start using this Day 1 when it is released.

File: app/ui/client/components/header/headerRoom.js [hunk 4]
```diff
@@ -120,7 +119,7 @@ Template.headerRoom.helpers({
 		const statusText = roomTypes.getUserStatusText(roomData.t, this._id);
 
 		if (s.trim(statusText)) {
-			return emojione.render(s.escapeHTML(statusText));
+			return statusText;
```
Comment (wreiske) [inline app/ui/client/components/header/headerRoom.js]: Is there a plan to add back support for rendering emojis in the status messages?

Comment (Hudell) [reply app/ui/client/components/header/headerRoom.js]: Yes, I just removed it now to make sure the PR is ready for the version cut, but I'll implement the new emoji renderer  once I'm done with some other work I'm busy with.

**Comment by wreiske** [comment]
🚀

**Comment by reetp** [comment]
Just been trying to change the status in open.rocket.chat but can't seem to do it reliably and cannot remove it to set back to default. Something is clearly broken in there.

Not sure whether to open a bug, add to  PR https://github.com/RocketChat/Rocket.Chat/pull/14853 or stay on this?

**Comment by Hudell** [comment]
Since the PRs have already been merged, I think opening a new issue would be ideal.

**Comment by wreiske** [comment]
This may be the issue:
https://github.com/RocketChat/Rocket.Chat/blob/24ebd2c4348a1b2f7f79038b89104f2866bf703a/app/lib/server/functions/setStatusText.js#L48

The change status server method is rate limited to once every 60 seconds.

**Comment by reetp** [comment]
As we discovered, it isn't obvious.
Make a change, hit save, realise you made a typo, modify and save, and it stays the same. And you have no idea why.

**Comment by wreiske** [comment]
@reetp if you wouldn't mind taking a minute to open up an issue, refencing that line, I'll get a pull request in to fix it. 

Or, you can make a pull request! I would say once every 2 or 3 seconds to account for edits. I was surprised there wasn't any notification about the rate limiter being hit. No "please slow down" messages or anything in console about it hitting the rate limit... hmm. If you open an issue we can discuss what an appropriate rate limit would be for it.

**Comment by reetp** [comment]
OK. Need some Zzzzz - Ill spin it up tomoz

**Pull request merged**